pkg/finality-grandpa/environment_test.go

```diff
@@ -0,0 +1,353 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/exp/rand"
+)
+
+type ID uint32
+
+type Signature uint32
+
+type timer struct {
+	wakerChan *wakerChan[error]
+	expired   bool
+}
+
+func newTimer(in <-chan time.Time) *timer {
+	inErr := make(chan error)
+	wc := newWakerChan(inErr)
+	t := timer{wakerChan: wc}
+	go func() {
+		<-in
+		inErr <- nil
+		t.expired = true
+	}()
+	return &t
+}
+
+func (t *timer) SetWaker(waker *waker) {
+	t.wakerChan.setWaker(waker)
+}
+
+func (t *timer) Elapsed() (bool, error) {
+	return t.expired, nil
```

Comment:
This could cause a race condition since `t.expired` is being written by a goroutine inside the `newTimer` function, and being read here, you can use an `atomic.Bool` (https://pkg.go.dev/sync/atomic#Bool) or a `sync.Mutex`

Reply:
This is test code, so I tried to keep it simple.  I know it's not going to panic with concurrent read/write, and this whole function gets polled again anyways.  I would definitely put a mutex around it if really implementing `Timer`.